### PR TITLE
Add HKDF key derivation function

### PR DESCRIPTION
**Pull request by GangSecurity**
Add HKDF key derivation support in ACRN based on mBedTLS crypto library

### CODE REVIEW

**Comment by NanlinXie** [comment]
Can one of the admins verify this patch?

File: hypervisor/lib/crypto/mbedtls/config.h [hunk 1]
```diff
@@ -2605,7 +2605,7 @@
  *            on it, and considering stronger message digests instead.
  *
  */
-#define MBEDTLS_SHA1_C
+//#define MBEDTLS_SHA1_C
```
Comment (bzhu5) [inline hypervisor/lib/crypto/mbedtls/config.h]: for coding style, should we use "//"?

Comment (GangSecurity) [reply hypervisor/lib/crypto/mbedtls/config.h]: The file is copied from mBedtls lib, "//" is the original coding style to comment code line in this file.

**Comment by lijinxia** [comment]
OK to verify

File: hypervisor/lib/crypto/mbedtls/config.h [hunk 2]
```diff
@@ -1000,7 +1000,7 @@
  *
  * Enable functions that use the filesystem.
  */
-#define MBEDTLS_FS_IO
+//#define MBEDTLS_FS_IO
```
Comment (anthonyzxu) [inline hypervisor/lib/crypto/mbedtls/config.h]: please remove this line

File: hypervisor/lib/crypto/mbedtls/md.h [hunk 3]
```diff
@@ -41,6 +41,8 @@
 #define MBEDTLS_ERR_MD_FILE_IO_ERROR                      -0x5200  /**< Opening or reading of file failed. */
 #define MBEDTLS_ERR_MD_HW_ACCEL_FAILED                    -0x5280  /**< MD hardware accelerator failed. */
 
+#define mbedtls_platform_zeroize(buf, len) memset(buf, 0, len)
+
 #ifdef __cplusplus
 extern "C" {
```
Comment (anthonyzxu) [inline hypervisor/lib/crypto/mbedtls/md.h]: it is always c in ACRN

File: hypervisor/lib/crypto/mbedtls/hkdf.c [hunk 4]
```diff
@@ -0,0 +1,190 @@
+/*
+ *  HKDF implementation -- RFC 5869
+ *
+ *  Copyright (C) 2016-2018, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
```
Comment (anthonyzxu) [inline hypervisor/lib/crypto/mbedtls/hkdf.c]: it is Apache license, any issue here?

File: hypervisor/lib/crypto/mbedtls/hkdf.h [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+ *  \return 0 on success.
+ *  \return #MBEDTLS_ERR_HKDF_BAD_INPUT_DATA when the parameters are invalid.
+ *  \return An MBEDTLS_ERR_MD_* error for errors returned from the underlying
+ *          MD layer.
+ */
+int mbedtls_hkdf_expand( const mbedtls_md_info_t *md, const unsigned char *prk,
+                         size_t prk_len, const unsigned char *info,
+                         size_t info_len, unsigned char *okm, size_t okm_len );
+
+#ifdef __cplusplus
```
Comment (anthonyzxu) [inline hypervisor/lib/crypto/mbedtls/hkdf.h]: it is always C in ACRN. do we need this?

Comment (dongyaozu) [reply hypervisor/lib/crypto/mbedtls/hkdf.h]: remove it

File: hypervisor/lib/crypto/mbedtls/config.h (hunk 1 above)
Comment (dongyaozu) [reply hypervisor/lib/crypto/mbedtls/config.h]: make sure check patch is fine.

File: hypervisor/lib/crypto/mbedtls/md_internal.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    void (*ctx_free_func)( void *ctx );
+
+    /** Clone state from a context */
+    void (*clone_func)( void *dst, const void *src );
+
+    /** Internal use only */
+    int (*process_func)( void *ctx, const unsigned char *input );
+};
+
+#if defined(MBEDTLS_MD2_C)
```
Comment (bzhu5) [inline hypervisor/lib/crypto/mbedtls/md_internal.h]: remove all but only MBEDTLS_SHA256_C

File: hypervisor/lib/crypto/mbedtls/ChangeLog [hunk 7]
```diff
@@ -0,0 +1,2884 @@
+mbed TLS ChangeLog (Sorted per branch, date)
```
Comment (bzhu5) [inline hypervisor/lib/crypto/mbedtls/ChangeLog]: keep it.

File: hypervisor/lib/crypto/mbedtls/hkdf.c (hunk 4 above)
Comment (bzhu5) [reply hypervisor/lib/crypto/mbedtls/hkdf.c]: we don't have issue

File: hypervisor/lib/crypto/mbedtls/md.c [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+#if defined(MBEDTLS_MD_C)
+
+#include "mbedtls/md.h"
+#include "mbedtls/md_internal.h"
+#include "mbedtls/platform_util.h"
+
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdlib.h>
```
Comment (bzhu5) [inline hypervisor/lib/crypto/mbedtls/md.c]: should remove

File: hypervisor/lib/crypto/mbedtls/hkdf.h [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+/**
+ *  \name HKDF Error codes
+ *  \{
+ */
+#define MBEDTLS_ERR_HKDF_BAD_INPUT_DATA  -0x5F80  /**< Bad input parameters to function. */
+/* \} name */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
```
Comment (dongyaozu) [inline hypervisor/lib/crypto/mbedtls/hkdf.h]: still have this..

File: hypervisor/lib/crypto/mbedtls/sha256.h [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+
+#include <stddef.h>
+#include <stdint.h>
+
+#define MBEDTLS_ERR_SHA256_HW_ACCEL_FAILED                -0x0037  /**< SHA-256 hardware accelerator failed */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
```
Comment (dongyaozu) [inline hypervisor/lib/crypto/mbedtls/sha256.h]: remove

File: hypervisor/lib/crypto/mbedtls/sha256.h [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+#if !defined(MBEDTLS_SHA256_ALT)
+// Regular implementation
+//
+
```
Comment (dongyaozu) [inline hypervisor/lib/crypto/mbedtls/sha256.h]: These 2 "//" ?

File: hypervisor/lib/crypto/mbedtls/sha256.h [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+    unsigned char buffer[64];   /*!< The data block being processed. */
+    int is224;                  /*!< Determines which function to use:
+                                     0: Use SHA-256, or 1: Use SHA-224. */
+}
+mbedtls_sha256_context;
+
+#else  /* MBEDTLS_SHA256_ALT */
+#include "sha256_alt.h"
+#endif /* MBEDTLS_SHA256_ALT */
+
```
Comment (dongyaozu) [inline hypervisor/lib/crypto/mbedtls/sha256.h]: Under what case, do we choose different option?  MBEDTLS_SHA256_ALT

File: hypervisor/lib/crypto/mbedtls/sha256.h [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+int mbedtls_internal_sha256_process( mbedtls_sha256_context *ctx,
+                                     const unsigned char data[64] );
+
+#if !defined(MBEDTLS_DEPRECATED_REMOVED)
+#if defined(MBEDTLS_DEPRECATED_WARNING)
+#define MBEDTLS_DEPRECATED      __attribute__((deprecated))
+#else
+#define MBEDTLS_DEPRECATED
+#endif
+/**
```
Comment (dongyaozu) [inline hypervisor/lib/crypto/mbedtls/sha256.h]: Can we simplify this?

File: hypervisor/lib/crypto/mbedtls/sha256.h [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+                        unsigned char output[32],
+                        int is224 );
+
+#if !defined(MBEDTLS_DEPRECATED_REMOVED)
+#if defined(MBEDTLS_DEPRECATED_WARNING)
+#define MBEDTLS_DEPRECATED      __attribute__((deprecated))
+#else
+#define MBEDTLS_DEPRECATED
+#endif
+
```
Comment (dongyaozu) [inline hypervisor/lib/crypto/mbedtls/sha256.h]: ditto

File: hypervisor/lib/crypto/mbedtls/sha256.h [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+ *
+ * \return         \c 0 on success.
+ * \return         \c 1 on failure.
+ */
+int mbedtls_sha256_self_test( int verbose );
+
+#ifdef __cplusplus
+}
+#endif
+
```
Comment (dongyaozu) [inline hypervisor/lib/crypto/mbedtls/sha256.h]: ditto

File: hypervisor/lib/crypto/mbedtls/sha256.h [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+ * \param output   The SHA-224 or SHA-256 checksum result.
+ * \param is224    Determines which function to use:
+ *                 0: Use SHA-256, or 1: Use SHA-224.
+ */
+MBEDTLS_DEPRECATED void mbedtls_sha256( const unsigned char *input,
+                                        size_t ilen,
+                                        unsigned char output[32],
+                                        int is224 );
+
+#undef MBEDTLS_DEPRECATED
```
Comment (dongyaozu) [inline hypervisor/lib/crypto/mbedtls/sha256.h]: ditto

File: hypervisor/lib/crypto/mbedtls/sha256.h (hunk 12 above)
Comment (GangSecurity) [reply hypervisor/lib/crypto/mbedtls/sha256.h]: MBEDTLS__MODULE_NAME__ALT:  Define it to let mbed TLS use your alternate core implementation of a symmetric crypto, an arithmetic or hash module.  We didn't implement SHA256 ourselves, so no other choice here.

File: hypervisor/lib/crypto/mbedtls/hkdf.h (hunk 9 above)
Comment (bzhu5) [reply hypervisor/lib/crypto/mbedtls/hkdf.h]: will remove this

File: hypervisor/lib/crypto/mbedtls/md.c [hunk 17]
```diff
@@ -0,0 +1,475 @@
+/**
+ * \file mbedtls_md.c
+ *
+ * \brief Generic message digest wrapper for mbed TLS
+ *
+ * \author Adriaan de Jong <dejong@fox-it.com>
+ *
+ *  Copyright (C) 2006-2015, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
```
Comment (bzhu5) [inline hypervisor/lib/crypto/mbedtls/md.c]: copyrigh (C) 2018, Intel....

**Comment by anthonyzxu** [comment]
MBEDTLS_MD_SHA224 is not used , can we remove it?
mbedtls_md_info_from_string is not used, can we remove it?
change function name from sha224_* to sha256_* in file mbedtls/md_wrap.c?

**Comment by bzhu5** [comment]
SHA224 code is gated by SHA2 flag, which is the same flag to control SHA256.

But yes, actually, we don’t use SHA224, but only SHA256 now.

@Gang, maybe we can remove that too to further reduce code lines

**Comment by GangSecurity** [comment]
Below 3 comments are solved.

MBEDTLS_MD_SHA224 is not used , can we remove it?
mbedtls_md_info_from_string is not used, can we remove it?
change function name from sha224_* to sha256_* in file mbedtls/md_wrap.c?

**Comment by bzhu5** [comment]
now looks good to me. but just change "HV: abstract hkdf key derivation files from mbedtls"
to "HV: **extract** hkdf key derivation files from mbedtls"

**Comment by lijinxia** [comment]
Verified+1

**Pull request merged**